solo-io/gloo
# Renew Hashicorp Vault Tokens

# Description

Implement token renewal as demonstrated here: https://github.com/hashicorp/vault-examples/blob/main/examples/token-renewal/go/example.go

## API changes
Added `settings.secretOptions.sources[].vault.aws.watcherIncrement` which gets used as the `Increment` argument when creating the `LifeTimewatcher` to monitor the token. If the value is 0 or not defined, the argument will not be passed.

## Code changes
Added renewal logic to `projects/gloo/pkg/bootstrap/clients/vault.go`

## Docs changes
Will be coming along

## Interesting decisions

-  Put the renewal logic in the `vault.go` bootstrap client in `gloo`  instead of in `solo-kit` because we are not currently handling any login/token/session logic in `solo-kit`
- Put the new `watcherIncrement` API/Helm field in the `VaultAwsAuth` section because it is part of the Vault/AWS connection

## Testing steps
Follow the steps in https://docs.solo.io/gloo-edge/main/guides/integrations/vault/aws-irsa-auth/ to see the 403 errors.

Repeat with the code in this branch and with the watcher increment set in settings (this shows only the new value to add, not the complete YAML block):
```
  secretOptions:
    sources:
    - vault:
        aws:
          watcherIncrement: 300
```

and the errors will not occur

## Notes for reviewers

Be sure to verify intended behavior by actually running it!

# Checklist:
